lib/modules/apostrophe-areas/public/js/user.js

```diff
@@ -148,11 +148,95 @@ apos.define('apostrophe-areas', {
         if ($el.attr('data-initialized')) {
           return;
         }
-        apos.create('apostrophe-areas-editor', { $el: $el, action: self.options.action });
+        var areaId = $el.attr('data-doc-id') + '.' + $el.attr('data-dot-path');
+        if (_.has(self.editors, areaId)) {
+          // If there is already an existing editor for this area, use that.
+          // Without this, inserting an image piece as part of editing a
+          // widget will destroy our ability to accept the widget's new
+          // value and save the area.
+          self.editors[areaId].resetEl($el);
+        } else {
+          apos.create('apostrophe-areas-editor', { $el: $el, action: self.options.action });
+        }
       });
       self.enhanceControlsHover(sel);
     };
 
+    self.register = function(docId, dotPath, editor) {
+      self.editors[docId + '.' + dotPath] = editor;
+    };
+
+    self.remapDotPaths = function() {
+      $('body').find('[data-apos-area]').each(function() {
+        var $area = $(this);
+        var docId = $area.attr('data-doc-id');
+        if (!docId) {
+          return;
+        }
+        if ($area.parents('[data-doc-id="' + docId + '"]').length) {
```

Comment:
The recalculate method we'll call recurses through descendants, so avoid those that will be reached that way.